tools/prow-auto-bumper/parser.go

```diff
@@ -0,0 +1,197 @@
+// and creates PRs updating them in knative/test-infra
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"math"
+	"strings"
+	"time"
+
+	"github.com/knative/test-infra/shared/ghutil"
+)
+
+// Helper method for adding a newly discovered tag into pv
+func (pv *PRVersions) getIndex(image, tag string) int {
+	if _, ok := pv.images[image]; !ok {
+		pv.images[image] = make([]versions, 0, 0)
+	}
+	_, variant := deconstructTag(tag)
+	iv := -1
+	for i, vs := range pv.images[image] {
+		if vs.variant == variant {
+			iv = i
+			break
+		}
+	}
+	if -1 == iv {
+		pv.images[image] = append(pv.images[image], versions{variant: variant})
+		iv = len(pv.images[image]) - 1
+	}
+	return iv
+}
+
+// Tags could be in the form of: v[YYYYMMDD]-[GIT_HASH](-[VARIANT_PART]),
+// separate it to `v[YYYYMMDD]-[GIT_HASH]` and `[VARIANT_PART]`
+func deconstructTag(in string) (string, string) {
+	dateCommit := in
+	var variant string
+	parts := strings.Split(in, "-")
+	if len(parts) > 2 {
+		variant = strings.Join(parts[2:], "-")
+	}
+	if len(parts) > 1 {
+		dateCommit = fmt.Sprintf("%s-%s", parts[0], parts[1])
+	}
+	return dateCommit, variant
+}
+
+// get key with highest value
+func getDominantKey(m map[string]int) string {
+	var res string
+	for key, v := range m {
+		if "" == res || v > m[res] {
+			res = key
+		}
+	}
+	return res
+}
+
+func (pv *PRVersions) getDominantVersions() versions {
+	if nil != pv.dominantVersions {
+		return *pv.dominantVersions
+	}
+
+	cOld := make(map[string]int)
+	cNew := make(map[string]int)
+	for _, vss := range pv.images {
+		for _, vs := range vss {
+			normOldTag, _ := deconstructTag(vs.oldVersion)
+			normNewTag, _ := deconstructTag(vs.newVersion)
+			cOld[normOldTag]++
+			cNew[normNewTag]++
+		}
+	}
+
+	pv.dominantVersions = &versions{
+		oldVersion: getDominantKey(cOld),
+		newVersion: getDominantKey(cNew),
+	}
+
+	return *pv.dominantVersions
+}
+
+// parse changelist, find all version changes, and store them in image name: versions map
+func (pv *PRVersions) parseChangelist(gcw *GHClientWrapper, gi gitInfo) error {
+	fs, err := gcw.ListFiles(gi.org, gi.repo, *pv.PR.Number)
+	if nil != err {
+		return err
+	}
+	for _, f := range fs {
+		if nil == f.Patch {
+			continue
+		}
+		minuses := imageMinusRegexp.FindAllStringSubmatch(*f.Patch, -1)
+		for _, minus := range minuses {
+			iv := pv.getIndex(minus[imageImagePart], minus[imageTagPart])
+			pv.images[minus[imageImagePart]][iv].oldVersion = minus[imageTagPart]
+		}
+
+		pluses := imagePlusRegexp.FindAllStringSubmatch(*f.Patch, -1)
+		for _, plus := range pluses {
+			iv := pv.getIndex(plus[imageImagePart], plus[imageTagPart])
+			pv.images[plus[imageImagePart]][iv].newVersion = plus[imageTagPart]
+		}
+	}
+
+	return nil
+}
+
+// Query all PRs from "k8s-ci-robot:autobump", find PR roughly 7 days old and was not reverted later.
+// Only return error if it's github related
+func getBestVersion(gcw *GHClientWrapper, gi gitInfo) (*PRVersions, error) {
```

Comment:
This function looks quite long... Can we possibly break it up?

Reply:
I have thought about it and tried it, it was not as clean as I thought, so kept it this way. Open to suggestion though

Reply:
I did not find a way to break it down either, but I do think there should be more comments. `getDominantVersions` does not have comments on itself or in this function call, and it's not quite clear why you need the check `if updatePR, ok := visited[vs.newVersion]; ok `.

Reply:
I was planning to add more in code README as well as a separate README file in another PR, but You're right, comments added here